lua/lib/lfospec.lua

```diff
@@ -0,0 +1,449 @@
+
+  if _lfo.enabled == 1 then
+
+    local current_val = (math.sin(2*math.pi*phase) + 1)/2
+    if _lfo.shape == "saw" then
+      current_val = phase < 0.5 and phase/0.5 or 1-(phase-0.5)/(0.5)
+    end
+
+    local min = _lfo.min
+    local max = _lfo.max
+
+    if min > max then
+      local old_min = min
+      local old_max = max
+      min = old_max
+      max = old_min
+    end
+
+    if min == -inf or max == -inf then
+      min = min == -inf and -2^1023.999999999 or min
+      max = max == inf and 2^1023.999999999 or max
+    end
+
+    local percentage = math.abs(min-max) * _lfo.depth
+    if _lfo.shape ~= 'random' then
+      _lfo.raw = current_val
+    end
+    current_val = current_val + _lfo.offset
+    local value = util.linlin(0,1,min,min + percentage,current_val)
+
+    if _lfo.depth > 0 then
+      local mid;
+      local scaled_min;
+      local scaled_max;
+      local raw_value;
+
+      if _lfo.baseline == 'min' then
+	scaled_min = min
+	scaled_max = min + percentage
+	mid = util.linlin(min,max,scaled_min,scaled_max,(min+max)/2)
+      elseif _lfo.baseline == 'center' then
+	mid = (min+max)/2
+	local centroid_mid = math.abs(min-max) * ((_lfo.depth)/2)
+	scaled_min = util.clamp(mid - centroid_mid,min,max)
+	scaled_max = util.clamp(mid + centroid_mid,min,max)
+	value = util.linlin(0,1,scaled_min,scaled_max,current_val)
+      elseif _lfo.baseline == 'max' then
+	mid = (min+max)/2
+	value = max - value
+	scaled_min = max * (1-(_lfo.depth))
+	scaled_max = max
+	mid = math.abs(util.linlin(min,max,scaled_min,scaled_max,mid))
+	value = util.linlin(0,1,scaled_max,scaled_min,current_val)
+      end
+
+      raw_value = util.linlin(min,max,0,1,value)
+
+      if _lfo.shape == "sine" then
+	value = util.clamp(value,min,max)
+	_lfo.scaled = value
+      elseif _lfo.shape == "saw" then
+	value = util.clamp(value,min,max)
+	_lfo.scaled = value
+      elseif _lfo.shape == "square" then
+	local square_value = value >= mid and max or min
+	square_value = util.linlin(min,max,scaled_min,scaled_max,square_value)
+	square_value = util.clamp(square_value,scaled_min,scaled_max)
+	_lfo.scaled = square_value
+	_lfo.raw = util.linlin(scaled_min,scaled_max,0,1,square_value)
+      elseif _lfo.shape == "random" then
+	local prev_value = rand_values
+	rand_values = value >= mid and max or min
+	local rand_value;
+	if prev_value ~= rand_values then
+	  rand_value = util.linlin(min,max,scaled_min,scaled_max,math.random(math.floor(min*100),math.floor(max*100))/100)
+	  rand_value = util.clamp(rand_value,min,max)
+	  _lfo.scaled = rand_value
+	  _lfo.raw = util.linlin(scaled_min,scaled_max,0,1,rand_value)
+	end
+      end
+      _lfo.action(_lfo.scaled, _lfo.raw)
+      if _lfo.parameter_id ~= nil then
+	params:set("lfo_scaled_".._lfo.parameter_id,util.round(_lfo.scaled,0.01))
+	params:set("lfo_raw_".._lfo.parameter_id,util.round(_lfo.raw,0.01))
+      end
+    end
+  end
+end
+
+hook['script_pre_init']:register("lfospec lattice setup", function()
```

Comment:
I do think using the mod hook system here is peculiar. I think explicit `init`/`cleanup` functions would be preferable.

honestly I also think I need to understand the rationale behind using lattice vs. just the clock system? what does that provide additionally?